### ISSUE TITLE
Add callback parameter to apply_kernel

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-processes#48]
In contrast to apply and apply_dimension, apply_kernel doesn't allow passing a callback such as a UDF. Nevertheless, this could be useful as mentioned by @appelmar today. So we could ass a process parameter, which then passes the data array according to the kernel specified (ignoring the actual array values of the kernel parameter).

**Comment by m-mohr** [comment]
Telco: Agreed to add a callback.

**Comment by m-mohr** [comment]
How would you expect the kernel data to be passed to the callback?

For example, let the kernel be `[[1, 1, 1], [1, 1, 1], [1, 1, 1]]` (so essentially a 3x3 kernel) for a datacube (x,y) what would you expect to be passed? How would you work with the data having the openEO processes in mind?
1. data = `[1, 2, 3, 3, 2, 1, 1, 2, 3]`
    sizeX = 3
    sizeY = 3
2. data = `[[1, 2, 3], [3, 2, 1], [1, 2, 3]]`

**Comment by soxofaan** [comment]
Option 2 (explicitly nested multidimensional arrays) seems to be most useful, unless I'm missing something

**Comment by m-mohr** [comment]
The biggest issue with option 2 is that it only works well with UDFs. We don't really have pre-defined processes that can work well with multi-dimensional arrays (unless people like messing around with array_element).

I'm personally not so much into kernel-based processing. How does this work in other tools?

**Comment by metzm** [comment]
There are a number of commonly used filter kernels, e.g. box, bartlett, gauss, normal, hermite, sinc, lanczos1, lanczos2, lanczos3, hann, hamming, blackman. How about discarding the data option and replacing it with a kernel option where one can select a predefined kernel?

About size: a more general option would be the radius of a filter kernel.

A somewhat more sophisticated option would be to combine different kernels with different radii, e.g. box with gauss, that allows to fine-tune the gauss kernel which is by nature infinite (radius to be interpreted as stddev of the gauss kernel). Of course users can calculate such kernel weights themselves, but it might be more convenient to be able to choose between different pre-defined and commonly used filter kernels.

**Comment by m-mohr** [comment]
> There are a number of commonly used filter kernels, e.g. box, bartlett, gauss, normal, hermite, sinc, lanczos1, lanczos2, lanczos3, hann, hamming, blackman. How about discarding the data option and replacing it with a kernel option where one can select a predefined kernel?

@matzm I'm not so much into all these filters filters, but there's already a kernel parameter, which allows to define kernels, and a multiplier parameter, too. Wouldn't that be enough already for all or some of the mentioned filters? I'm pretty sure gauss is covered...

The callback is meant to allow mor advanced filter opterations, for example in a UDF.

**Comment by metzm** [comment]
> @metzm Thank you. I'm not so much into all these filters filters, but there's already a kernel parameter, which allows to define kernels, and a multiplier parameter, too (see [apply_kernel](https://processes.openeo.org/#apply_kernel)). Wouldn't that be enough already for all or some of the mentioned filters? I'm pretty sure gauss is covered...

The kernel parameter is of type array, usually a multi-dimensional array of numbers. My point is that it is more user-friendly to provide pre-defined kernel filters which calculate such an array of numbers internally.

The current multiplier, parameter factor with default value 1, is multiplied to each value computed by the focal operation. If this factor is not 1, results are no longer comparable to the input. Is this intended?

gauss, as any other filter kernel, is currently covered as long as the corresponding array of numbers is passed to the kernel parameter. IMHO, this is not very user-friendly. What is easier: let the user specify e.g. kernel=gauss or let the user calculate the corresponding array first, then provide it to the kernel parameter as multi-dimensional array of numbers?

I guess that users will decide on a suitable kernel filter based on some recommendations from literature and would like to apply this filter without the need to bother about the maths behind that filter. 
> 
> The callback is meant to allow mor advanced filter opterations, for example in a UDF.

Can you provide an example?

**Comment by m-mohr** [comment]
> The kernel parameter is of type array, usually a multi-dimensional array of numbers. My point is that it is more user-friendly to provide pre-defined kernel filters which calculate such an array of numbers internally. [...] gauss, as any other filter kernel, is currently covered as long as the corresponding array of numbers is passed to the kernel parameter. IMHO, this is not very user-friendly. What is easier: let the user specify e.g. kernel=gauss or let the user calculate the corresponding array first, then provide it to the kernel parameter as multi-dimensional array of numbers?

Ah, now I get the point. The issue is like with all "pre-defined lists" that it may be too much or not enough, depending on the implementor. Your are limiting what users can use, but on the other hand it's easier to use. Maybe we can pre-define some common ones. We could also define processes that simply return multi-dimensional kernels, maybe this is easier to extend and limit. Let's discuss this on one of the telcos with all participants.

> The current multiplier, parameter factor with default value 1, is multiplied to each value computed by the focal operation. If this factor is not 1, results are no longer comparable to the input. Is this intended?

I don't get this, but as I said I'm not really into filters etc.

> Can you provide an example?

No, i did not propose this. Maybe someone can assist? @jdries @lforesta or so?

**Comment by jdries** [comment]
You can have a look at this example to see how easy it is to apply a gaussian kernel:
https://github.com/Open-EO/openeo-processes/blob/uc-phenology/examples/phenology-uc2/Phenology.ipynb
More specifically:
```
def makekernel(iwindowsize):            
            kernel_vect = scipy.signal.windows.gaussian(iwindowsize, std = iwindowsize/4.0, sym=True)
            kernel = np.outer(kernel_vect, kernel_vect)
            kernel = kernel / kernel.sum()            
            return kernel   
```
The python (and probably R) ecosystems provide sufficient convenience to easily define a kernel. Adding predefined kernels to openeo would require a lot of additional api to be defined, compared to what we gain.
I don't entirely disagree that this kind of feature would be more user friendly, but there's (a lot of) other things to deal with before it becomes relevant enough IMO. (Even allowing a callback is something we might not implement in the near future.)

**Comment by soxofaan** [comment]
The python client could  provide some helper functions for quick and easy building of these popular kernels  -> Open-EO/openeo-python-client#106

**Comment by m-mohr** [comment]
Agreed. We can solve this either on the client-side (based on the subtype kernel) or back-ends can simply implement something like gaussian_window as process, which returns a multi-dimensional array that can then be passed to the kernel-parameter. In this way we don't need to add it to apply_kernel directly.

My question from https://github.com/Open-EO/openeo-processes/issues/48#issuecomment-566125654 is still open though...

**Comment by metzm** [comment]
I still think that it would be a simplification of the API and more fail-safe if the apply_kernel process does not accept data, but instead the name of a filter kernel (as string, not separate process) and a radius. It is fairly easy in backends to implement these filter kernels, and some backends might already have these kernels implemented. This is an attempt of an answer to [#48 (comment)](https://github.com/Open-EO/openeo-processes/issues/48#issuecomment-566125654l).

**Comment by soxofaan** [comment]
> it would be a simplification of the API and more fail-safe if the apply_kernel process does not accept data

While I agree that specifying a kernel with a common name as string could be a user friendly addition to the API, I don't think that you can just drop the option to specify a kernel as a user-specified array, quite some end users are going to want to use non-standard, custom kernels. 

As noted above, even if the kernels at API level are array-only, it is still possible (and probably easier/more flexible) to implement this "kernel as a string" feature at the level of the openEO client.

**Comment by aljacob** [comment]
I agree with Stefaan. Different pre-defined standard kernels put be put easily put into the clients and just automatically fill in the correct array for the process call then.

**Comment by m-mohr** [comment]
Telco: Seems there's no use case at the moment. We'll revisit this later. One idea also was to define an additional process for it and not put it into the current apply_kernel process.
